pkg/registry/generic/etcd/etcd.go

```diff
@@ -455,3 +444,18 @@ func (e *Etcd) WatchPredicate(ctx api.Context, m generic.Matcher, resourceVersio
 
 	return e.Helper.WatchList(e.KeyRootFunc(ctx), version, filterFunc)
 }
+
+// calculateTTL is a helper for retrieving the updated TTL for an object or returning an error
+// if the TTL cannot be calculated. The defaultTTL is changed to zero if less than zero.
+func (e *Etcd) calculateTTL(obj runtime.Object, defaultTTL int64, update bool) (ttl uint64, err error) {
+	// etcd may return a negative TTL for a node if the expiration has not occured due
+	// to server lag.
```

Comment:
Uh. Good point.  Actually, that means that you could push a deletion forever by making updates in the last second of an etcd object's life.  If we delete to 1, then at least there's a chance it gets deleted.  Seems like it has to be 1.

----- Original Message -----
> > @@ -455,3 +444,18 @@ func (e *Etcd) WatchPredicate(ctx api.Context, m
> > generic.Matcher, resourceVersio
> >  
> >  	return e.Helper.WatchList(e.KeyRootFunc(ctx), version, filterFunc)
> >  }
> > +
> > +// calculateTTL is a helper for retrieving the updated TTL for an object
> > or returning an error
> > +// if the TTL cannot be calculated. The defaultTTL is changed to zero if
> > less than zero.
> > +func (e *Etcd) calculateTTL(obj runtime.Object, defaultTTL int64, update
> > bool) (ttl uint64, err error) {
> > +	// etcd may return a negative TTL for a node if the expiration has not
> > occured due
> > +	// to server lag.
> 
> ... then it seems pretty bad to set it to 0 in that case, no? Isn't that
> going to permanently un-delete an object that should be deleted?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/8724/files#r31065663